### PR TITLE
Code generate proper benchmarks

### PR DESCRIPTION
**Pull request by andyscott**
Adds a benchmark suite to compare Iota to other coproduct implementations.

Here are preliminary results (against cats) with 1 to 10 algebras in the coproduct.
```
Benchmark         Mode  Cnt        Score         Error  Units
IotaBench.catsA  thrpt    5  5351847.767 ±  516985.216  ops/s
IotaBench.catsB  thrpt    5  5021697.068 ±  370027.385  ops/s
IotaBench.catsC  thrpt    5  4793059.293 ±  402162.692  ops/s
IotaBench.catsD  thrpt    5  4519549.507 ±  298976.147  ops/s
IotaBench.catsE  thrpt    5  4350670.585 ±  357685.493  ops/s
IotaBench.catsF  thrpt    5  4363635.968 ±   85969.402  ops/s
IotaBench.catsG  thrpt    5  4087097.327 ± 1194185.852  ops/s
IotaBench.catsH  thrpt    5  4024149.196 ±  171135.642  ops/s
IotaBench.catsI  thrpt    5  3858659.309 ±  135996.181  ops/s
IotaBench.catsJ  thrpt    5  3722032.947 ±  129992.126  ops/s
IotaBench.iotaA  thrpt    5  5438446.214 ±  612423.539  ops/s
IotaBench.iotaB  thrpt    5  5515331.980 ±  168528.822  ops/s
IotaBench.iotaC  thrpt    5  5443160.606 ±  898961.026  ops/s
IotaBench.iotaD  thrpt    5  5462806.255 ±  280228.707  ops/s
IotaBench.iotaE  thrpt    5  5406944.411 ±  522615.445  ops/s
IotaBench.iotaF  thrpt    5  5456559.008 ±  393859.771  ops/s
IotaBench.iotaG  thrpt    5  5423384.542 ±  501750.405  ops/s
IotaBench.iotaH  thrpt    5  5548054.140 ±  508072.479  ops/s
IotaBench.iotaI  thrpt    5  5503244.641 ±  556508.649  ops/s
IotaBench.iotaJ  thrpt    5  5487354.981 ±  250609.275  ops/s
```

Here's a plot of the results of the benchmark up to 25 algebras:

bench

And here's a dump of the generated benchmark code for 10 algebras: https://gist.github.com/andyscott/ff79b75d8e07e8336cf2ad33462133df

Note: This is an "uber" commit that also includes:
- a fix a bug I discovered along the way
- a new debugging implicit

### CODE REVIEW

**Comment by andyscott** [comment]
Here's a dump of generated benchmark code, for reference:
https://gist.github.com/andyscott/ff79b75d8e07e8336cf2ad33462133df

File: README.md [hunk 1]
```diff
@@ -1,4 +1,3 @@
-
```
Comment (juanpedromoreno) [inline README.md]: This line is needed for the auto-replacement feature

File: modules/readme/src/main/tut/README.md [hunk 2]
```diff
@@ -172,4 +185,6 @@ Iota is designed and developed by 47 Degrees
 
 Copyright (C) 2016-2017 47 Degrees. <http://47deg.com>
 
-[comment]: # (End Copyright)
\ No newline at end of file
+[comment]: # (End Copyright)
```
Comment (juanpedromoreno) [inline modules/readme/src/main/tut/README.md]: I'd suggest to leave the copyright notice at the end of the file.

File: modules/readme/src/main/tut/README.md [hunk 3]
```diff
@@ -1,4 +1,3 @@
-
```
Comment (juanpedromoreno) [inline modules/readme/src/main/tut/README.md]: This extra line is needed.

File: README.md (hunk 1 above)
Comment (andyscott) [inline README.md]: Gah! My emacs setup auto removes leading new lines!

File: modules/readme/src/main/tut/README.md (hunk 2 above)
Comment (andyscott) [inline modules/readme/src/main/tut/README.md]: It still is-- that's just an invisible link reference that's used above.

**Pull request merged**